packages/core-data/src/private-selectors.ts

```diff
@@ -105,3 +106,129 @@ export function getEntityRecordPermissions(
 export function getRegisteredPostMeta( state: State, postType: string ) {
 	return state.registeredPostMeta?.[ postType ] ?? {};
 }
+
+interface SiteData {
+	show_on_front?: string;
+	page_on_front?: string | number;
+	page_for_posts?: string | number;
+}
+
+export const getHomePage = createRegistrySelector( ( select ) => () => {
+	const siteData = select( STORE_NAME ).getEntityRecord( 'root', 'site' ) as
+		| SiteData
+		| undefined;
+	const homepageId =
+		siteData?.show_on_front === 'page' &&
+		[ 'number', 'string' ].includes( typeof siteData.page_on_front ) &&
+		siteData.page_on_front &&
+		!! +siteData.page_on_front // We also need to check if it's not zero(`0`).
+			? siteData.page_on_front.toString()
+			: null;
+	const isSiteLoaded = !! siteData;
+	if ( isSiteLoaded && homepageId ) {
+		return { postType: 'page', postId: homepageId };
+	}
+	if ( isSiteLoaded && ! homepageId ) {
+		const frontPageTemplateId = select( STORE_NAME ).getDefaultTemplateId( {
+			slug: 'front-page',
+		} );
+		return { postType: 'wp_template', postId: frontPageTemplateId };
+	}
+	return null;
+} );
+
+export const getPostsPageId = createRegistrySelector( ( select ) => () => {
+	const siteData = select( STORE_NAME ).getEntityRecord( 'root', 'site' ) as
+		| SiteData
+		| undefined;
+	return siteData?.show_on_front === 'page' &&
+		siteData.page_for_posts &&
+		[ 'number', 'string' ].includes( typeof siteData.page_for_posts )
+		? siteData.page_for_posts.toString()
+		: null;
+} );
+
+export const getTemplateId = createRegistrySelector(
+	( select ) => ( state, postType, postId ) => {
+		const homepage = unlock( select( STORE_NAME ) ).getHomePage();
+
+		// For the front page, we always use the front page template if existing.
+		if (
+			postType === 'page' &&
+			postType === homepage?.postType &&
+			postId === homepage?.postId
+		) {
```

Comment:
In this code, are you sufficiently taking into account the state where `homepage` is not yet loaded and therefore `undefined`? The old code had the `hasLoadedAllDependencies` flag and an early return. The new code will proceed to resolving the page template. The old code would do that only after being sure that the homepage info is loaded and really doesn't match.